bgpd/bgp_mplsvpn.c

```diff
@@ -1350,6 +1349,187 @@ leak_update(struct bgp *to_bgp, struct bgp_dest *bn,
 	return new;
 }
 
+static void
+bgp_mplsvpn_path_nh_label_map(struct bgp_path_info *path,
+			      struct bgp_label_per_nexthop_cache *blnc,
+			      bool make)
+{
+	if (path->label_nexthop) {
+		LIST_REMOVE(path, label_nh_thread);
+		path->label_nexthop->path_count--;
+		path->label_nexthop = NULL;
+	}
+	if (make) {
+		LIST_INSERT_HEAD(&(blnc->paths), path, label_nh_thread);
+		path->label_nexthop = blnc;
+		path->label_nexthop->path_count++;
+	}
+	if (blnc)
+		blnc->last_update = monotime(NULL);
+}
+
+void bgp_mplsvpn_path_nh_label_unlink(struct bgp_path_info *path)
+{
+	struct bgp_label_per_nexthop_cache *blnc = path->label_nexthop;
+
+	if (!blnc)
+		return;
+
+	bgp_mplsvpn_path_nh_label_map(path, NULL, false);
+	if (LIST_EMPTY(&(blnc->paths)))
+		bgp_label_per_nexthop_free(blnc);
+}
+static void
+bgp_mplsvpn_path_nh_label_update(struct bgp_label_per_nexthop_cache *blnc)
+{
+	struct bgp_table *table;
+	struct bgp_dest *dest;
+	struct bgp *bgp;
+	struct bgp_path_info *path;
+
+	LIST_FOREACH (path, &(blnc->paths), label_nh_thread) {
+		dest = path->net;
+		if (!dest)
+			continue;
+		table = bgp_dest_table(dest);
+		if (!table)
+			continue;
+		bgp = table->bgp;
+		vpn_leak_from_vrf_update(blnc->to_bgp, bgp, path);
+	}
+}
+
+static int bgp_mplsvpn_get_label_cb(mpls_label_t label, void *context,
+				    bool allocated)
+{
+	struct bgp_label_per_nexthop_cache *blnc =
+		(struct bgp_label_per_nexthop_cache *)context;
+	mpls_label_t old_label = blnc->label;
```

Comment:
Don't we need to check for NULL before dereferencing?

Reply:
ok